src/storage/patch/SparqlUpdatePatchHandler.ts

```diff
@@ -100,13 +129,42 @@ export class SparqlUpdatePatchHandler extends PatchHandler {
       this.logger.debug(`Patching new resource ${identifier.path}.`);
     }
 
-    // Apply the patch
-    store.removeQuads(deletes);
-    store.addQuads(inserts);
-    this.logger.debug(`Removed ${deletes.length} and added ${inserts.length} quads to ${identifier.path}.`);
+    this.applyOperation(store, op);
     this.logger.debug(`${store.size} quads will be stored to ${identifier.path}.`);
 
     // Write the result
     await this.source.setRepresentation(identifier, new BasicRepresentation(store.match() as Readable, INTERNAL_QUADS));
   }
+
+  /**
+   * Apply the given algebra update operation to the store of quads.
+   */
+  private applyOperation(store: Store<BaseQuad>, op: Algebra.Operation): void {
+    if (this.isDeleteInsert(op)) {
+      this.applyDeleteInsert(store, op);
+    // Only other options is Composite after passing `validateUpdate`
+    } else {
+      this.applyComposite(store, op as Algebra.CompositeUpdate);
+    }
+  }
+
+  /**
+   * Apply the given composite update operation to the store of quads.
+   */
+  private applyComposite(store: Store<BaseQuad>, op: Algebra.CompositeUpdate): void {
+    for (const update of op.updates) {
+      this.applyOperation(store, update);
+    }
+  }
+
+  /**
+   * Apply the given DELETE/INSERT update operation to the store of quads.
+   */
+  private applyDeleteInsert(store: Store<BaseQuad>, op: Algebra.DeleteInsert): void {
+    const deletes = op.delete ?? [];
+    const inserts = op.insert ?? [];
+    store.removeQuads(deletes);
+    store.addQuads(inserts);
+    this.logger.debug(`Removed ${deletes.length} and added ${inserts.length} quads.`);
```

Comment:
Is that true though? That's the mission you set out to do—and this should perhaps be logged _before_ doing it. But depending on the contents of the store, the results can be different.

Reply:
True. I just copy/pasted the previous logger statement but there is no guarantee these numbers are what actually happens.